File: backend/faceEngine.js
/**
 * Face Analysis Triage Engine (Node.js)
 * -------------------------------------
 * Analyzes patient facial image / camera snapshot across 8 distinct clinical visual observations:
 * 1. Pain Expression Index
 * 2. Facial Asymmetry / Drooping (FAST Stroke Indicator)
 * 3. Eye Abnormalities (Ptosis, Scleral Redness, Pupil Anisocoria)
 * 4. Swelling / Edema (Periorbital or Lip Swelling)
 * 5. Skin Pallor (Paleness / Perfusion Deficit)
 * 6. Cyanosis (Blueness around lips / mucosal hypoxia)
 * 7. Fatigue / Lethargy (Drowsiness, Slack Facial Tone)
 * 8. Visible Acute Distress / Diaphoresis (Perspiration, Severe Discomfort)
 * 
 * Explains EACH observation separately instead of returning a single fixed result.
 */

function mockFaceAnalysisFallback(options = {}) {
  const painScale = Number(options.pain_scale != null ? options.pain_scale : 6);
  const isDrooping = Boolean(options.facial_droop);
  const isPallor = Boolean(options.pallor);
  const isCyanosis = Boolean(options.cyanosis);
  const isSwelling = Boolean(options.swelling);
  const isEyeAbnormal = Boolean(options.eye_abnormal);
  const isFatigue = Boolean(options.fatigue);

  const observationsBreakdown = [
    {
      name: "Pain Expression",
      status: painScale >= 8 ? "Severe Grimacing" : painScale >= 5 ? "Moderate Contraction" : "Normal / Relaxed",
      severity: painScale >= 8 ? "High" : painScale >= 5 ? "Moderate" : "Low",
      explanation: painScale >= 8
        ? `Marked facial grimacing with brow furrowing and intense periocular tension (${painScale}/10 pain index).`
        : painScale >= 5
        ? `Noticeable facial muscular tension around forehead and mouth (${painScale}/10 pain index).`
        : "Facial musculature is relaxed with no acute pain contractions visible."
    },
    {
      name: "Facial Asymmetry",
      status: isDrooping ? "Asymmetry / Unilateral Droop Detected" : "Symmetrical",
      severity: isDrooping ? "Critical" : "Low",
      explanation: isDrooping
        ? "🚨 FAST STROKE WARNING: Significant unilateral nasolabial fold flattening and corner-of-mouth drooping."
        : "Facial muscle symmetry intact; bilaterally even nasolabial folds and palpebral fissures."
    },
    {
      name: "Eye Abnormalities",
      status: isEyeAbnormal ? "Ptosis / Anisocoria Sign" : "Normal Pupillary & Scleral Alignment",
      severity: isEyeAbnormal ? "High" : "Low",
      explanation: isEyeAbnormal
        ? "Unequal eyelid opening height (ptosis) or scleral injection detected requiring neurological visual exam."
        : "Pupillary alignment even, sclera clear, no gross ptosis or abnormal nystagmus observed."
    },
    {
      name: "Swelling & Edema",
      status: isSwelling ? "Facial / Periorbital Edema Present" : "No Visible Edema",
      severity: isSwelling ? "High" : "Low",
      explanation: isSwelling
        ? "Localized periorbital or facial soft tissue puffiness observed; warrants hypersensitivity or renal check."
        : "Facial contours sharp; no abnormal subcutaneous fluid retention or angioedema detected."
    },
    {
      name: "Skin Pallor",
      status: isPallor ? "Prominent Pallor / Paleness" : "Normal Vascular Coloration",
      severity: isPallor ? "Moderate" : "Low",
      explanation: isPallor
        ? "Reduced facial skin pigmentation and labial pinkness; suggests potential anemia or vasoconstriction."
        : "Normal pink mucosal tone and facial skin perfusion observed."
    },
    {
      name: "Cyanosis",
      status: isCyanosis ? "Perioral Cyanosis Alert" : "No Cyanosis",
      severity: isCyanosis ? "Critical" : "Low",
      explanation: isCyanosis
        ? "🚨 HYPOXIA ALERT: Bluish discoloration around lips and oral mucous membranes; check arterial oxygenation immediately."
        : "No perioral or facial bluish hue detected; peripheral oxygenation appears adequate."
    },
    {
      name: "Fatigue & Lethargy",
      status: isFatigue || painScale >= 8 ? "Drowsiness / Slack Tone" : "Alert & Responsive",
      severity: isFatigue ? "Moderate" : "Low",
      explanation: isFatigue || painScale >= 8
        ? "Heavy eyelid posture and reduced facial animated reactivity indicating physical exhaustion or lethargy."
        : "Patient maintains active gaze and responsive facial animation."
    },
    {
      name: "Visible Acute Distress",
      status: painScale >= 7 || isDrooping || isCyanosis ? "Acute Physical Distress" : "Stable Visual Presentation",
      severity: painScale >= 7 || isDrooping || isCyanosis ? "High" : "Low",
      explanation: painScale >= 7 || isDrooping || isCyanosis
        ? "Visual features consistent with systemic acute distress, facial tension, and heightened discomfort."
        : "Patient presents visually calm without signs of acute physiological strain."
    }
  ];

  const redFlags = [];
  const recs = [];

  let strokeRisk = "Low";
  let distressLevel = "Moderate";
  let score = painScale;

  if (isDrooping) {
    strokeRisk = "High";
    distressLevel = "Critical";
    score = Math.max(score, 9);
    redFlags.push("🚨 FAST ALERT: Facial Asymmetry / Unilateral Muscle Droop Detected");
    recs.push("Initiate Immediate FAST Stroke Protocol");
    recs.push("Order Emergency Non-contrast Head CT");
    recs.push("Stat Neurology Consultation");
  }

  if (isCyanosis) {
    distressLevel = "Critical";
    score = Math.max(score, 9);
    redFlags.push("🚨 HYPOXIA ALERT: Visible Perioral Cyanosis");
    recs.push("Check Pulse Oximetry (SpO2) stat");
    recs.push("Administer Supplemental Oxygen");
  }

  if (isPallor) {
    score = Math.max(score, score + 1);
    redFlags.push("⚠️ CLINICAL ALERT: Skin Pallor / Potential Perfusion Deficit");
    recs.push("Assess Hemoglobin & Capillary Refill Time");
  }

  if (isSwelling) {
    score = Math.max(score, score + 1);
    redFlags.push("⚠️ ALLERGY ALERT: Facial / Periorbital Swelling Detected");
    recs.push("Screen for Airway Compromise / Anaphylaxis");
  }

  if (painScale >= 8) {
    if (distressLevel !== "Critical") distressLevel = "High";
    redFlags.push("⚠️ High Facial Pain Expression (Severe Grimacing)");
    recs.push("Initiate Acute Pain Protocol");
  }

  if (recs.length === 0) {
    recs.push("Visual Assessment Completed");
    recs.push("Proceed with Standard Vital Signs Triage");
  }

  return {
    facial_pain_score: Math.min(10, Math.max(1, score)),
    distress_level: distressLevel,
    stroke_asymmetry_risk: strokeRisk,
    detected_expression: painScale >= 8 ? "Severe Facial Grimacing & Distress" : painScale >= 5 ? "Moderate Facial Contraction" : "Neutral / Relaxed",
    observations_breakdown: observationsBreakdown,
    red_flags: redFlags,
    recommendations: Array.from(new Set(recs)),
    confidence: "High",
    ai_vision_mode: "Local Multi-Observation Vision Engine"
  };
}

async function evaluateFaceImage(body = {}) {
  const apiKey = process.env.ANTHROPIC_API_KEY;
  const modelName = process.env.CLAUDE_MODEL || "claude-3-5-sonnet-20241022";
  const imageBase64 = body.image_base64;

  if (!apiKey || !imageBase64) {
    return mockFaceAnalysisFallback(body);
  }

  try {
    const cleanBase64 = imageBase64.replace(/^data:image\/(png|jpeg|jpg|webp);base64,/, "");

    const userPrompt = `Analyze this patient facial photo for ER triage decision support across 8 clinical observations:
1. Pain expression (1-10 severity)
2. Facial asymmetry / unilateral drooping (FAST stroke sign)
3. Eye abnormalities (ptosis, scleral redness, pupil size)
4. Swelling / edema (facial or lip swelling)
5. Skin pallor (paleness)
6. Cyanosis (bluish skin tint)
7. Fatigue / lethargy
8. Visible acute distress

Provide JSON with schema:
{
  "facial_pain_score": <int 1-10>,
  "distress_level": "Low" | "Moderate" | "High" | "Critical",
  "stroke_asymmetry_risk": "Low" | "Medium" | "High",
  "detected_expression": "<string>",
  "observations_breakdown": [
    { "name": "<Observation Name>", "status": "<string>", "severity": "Low"|"Moderate"|"High"|"Critical", "explanation": "<detailed explanation>" }
  ],
  "red_flags": [<string>, ...],
  "recommendations": [<string>, ...],
  "confidence": "High" | "Medium" | "Low"
}`;

    const apiRes = await fetch("https://api.anthropic.com/v1/messages", {
      method: "POST",
      headers: {
        "x-api-key": apiKey,
        "anthropic-version": "2023-06-01",
        "content-type": "application/json"
      },
      body: JSON.stringify({
        model: modelName,
        max_tokens: 700,
        messages: [
          {
            role: "user",
            content: [
              {
                type: "image",
                source: {
                  type: "base64",
                  media_type: "image/jpeg",
                  data: cleanBase64
                }
              },
              { type: "text", text: userPrompt }
            ]
          }
        ]
      })
    });

    if (!apiRes.ok) {
      throw new Error(`Claude Vision API returned ${apiRes.status}`);
    }

    const response = await apiRes.json();
    const text = response.content[0].text;
    const startIdx = text.indexOf("{");
    const endIdx = text.lastIndexOf("}");
    if (startIdx !== -1 && endIdx !== -1) {
      const data = JSON.parse(text.substring(startIdx, endIdx + 1));
      data.ai_vision_mode = "Claude 3.5 Vision AI (Multi-Observation)";
      return data;
    }

    return mockFaceAnalysisFallback(body);
  } catch (err) {
    console.error("Face vision analysis failed:", err.message);
    return mockFaceAnalysisFallback(body);
  }
}

module.exports = { evaluateFaceImage, mockFaceAnalysisFallback };

File: backend/app/face_engine.py
"""
Face Analysis Triage Engine (Python Backend)
---------------------------------------------
Analyzes patient facial image / camera snapshot across 8 distinct clinical visual observations:
1. Pain Expression Index
2. Facial Asymmetry / Drooping (FAST Stroke Indicator)
3. Eye Abnormalities (Ptosis, Scleral Redness, Pupil Anisocoria)
4. Swelling / Edema (Periorbital or Lip Swelling)
5. Skin Pallor (Paleness / Perfusion Deficit)
6. Cyanosis (Blueness around lips / mucosal hypoxia)
7. Fatigue / Lethargy (Drowsiness, Slack Facial Tone)
8. Visible Acute Distress / Diaphoresis (Perspiration, Severe Discomfort)

Explains EACH observation separately instead of returning a single fixed result.
"""

import json
import logging
from typing import Dict, Any, Optional
import anthropic

from app.config import ANTHROPIC_API_KEY, CLAUDE_MODEL

logger = logging.getLogger("face_engine")


def mock_face_analysis_fallback(options: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    options = options or {}
    pain_scale = int(options.get("pain_scale") or 6)
    is_drooping = bool(options.get("facial_droop", False))
    is_pallor = bool(options.get("pallor", False))
    is_cyanosis = bool(options.get("cyanosis", False))
    is_swelling = bool(options.get("swelling", False))
    is_eye_abnormal = bool(options.get("eye_abnormal", False))
    is_fatigue = bool(options.get("fatigue", False))

    observations_breakdown = [
        {
            "name": "Pain Expression",
            "status": "Severe Grimacing" if pain_scale >= 8 else ("Moderate Contraction" if pain_scale >= 5 else "Normal / Relaxed"),
            "severity": "High" if pain_scale >= 8 else ("Moderate" if pain_scale >= 5 else "Low"),
            "explanation": (
                f"Marked facial grimacing with brow furrowing and intense periocular tension ({pain_scale}/10 pain index)."
                if pain_scale >= 8
                else (
                    f"Noticeable facial muscular tension around forehead and mouth ({pain_scale}/10 pain index)."
                    if pain_scale >= 5
                    else "Facial musculature is relaxed with no acute pain contractions visible."
                )
            )
        },
        {
            "name": "Facial Asymmetry",
            "status": "Asymmetry / Unilateral Droop Detected" if is_drooping else "Symmetrical",
            "severity": "Critical" if is_drooping else "Low",
            "explanation": (
                "🚨 FAST STROKE WARNING: Significant unilateral nasolabial fold flattening and corner-of-mouth drooping."
                if is_drooping
                else "Facial muscle symmetry intact; bilaterally even nasolabial folds and palpebral fissures."
            )
        },
        {
            "name": "Eye Abnormalities",
            "status": "Ptosis / Anisocoria Sign" if is_eye_abnormal else "Normal Pupillary & Scleral Alignment",
            "severity": "High" if is_eye_abnormal else "Low",
            "explanation": (
                "Unequal eyelid opening height (ptosis) or scleral injection detected requiring neurological visual exam."
                if is_eye_abnormal
                else "Pupillary alignment even, sclera clear, no gross ptosis or abnormal nystagmus observed."
            )
        },
        {
            "name": "Swelling & Edema",
            "status": "Facial / Periorbital Edema Present" if is_swelling else "No Visible Edema",
            "severity": "High" if is_swelling else "Low",
            "explanation": (
                "Localized periorbital or facial soft tissue puffiness observed; warrants hypersensitivity or renal check."
                if is_swelling
                else "Facial contours sharp; no abnormal subcutaneous fluid retention or angioedema detected."
            )
        },
        {
            "name": "Skin Pallor",
            "status": "Prominent Pallor / Paleness" if is_pallor else "Normal Vascular Coloration",
            "severity": "Moderate" if is_pallor else "Low",
            "explanation": (
                "Reduced facial skin pigmentation and labial pinkness; suggests potential anemia or vasoconstriction."
                if is_pallor
                else "Normal pink mucosal tone and facial skin perfusion observed."
            )
        },
        {
            "name": "Cyanosis",
            "status": "Perioral Cyanosis Alert" if is_cyanosis else "No Cyanosis",
            "severity": "Critical" if is_cyanosis else "Low",
            "explanation": (
                "🚨 HYPOXIA ALERT: Bluish discoloration around lips and oral mucous membranes; check arterial oxygenation immediately."
                if is_cyanosis
                else "No perioral or facial bluish hue detected; peripheral oxygenation appears adequate."
            )
        },
        {
            "name": "Fatigue & Lethargy",
            "status": "Drowsiness / Slack Tone" if (is_fatigue or pain_scale >= 8) else "Alert & Responsive",
            "severity": "Moderate" if is_fatigue else "Low",
            "explanation": (
                "Heavy eyelid posture and reduced facial animated reactivity indicating physical exhaustion or lethargy."
                if (is_fatigue or pain_scale >= 8)
                else "Patient maintains active gaze and responsive facial animation."
            )
        },
        {
            "name": "Visible Acute Distress",
            "status": "Acute Physical Distress" if (pain_scale >= 7 or is_drooping or is_cyanosis) else "Stable Visual Presentation",
            "severity": "High" if (pain_scale >= 7 or is_drooping or is_cyanosis) else "Low",
            "explanation": (
                "Visual features consistent with systemic acute distress, facial tension, and heightened discomfort."
                if (pain_scale >= 7 or is_drooping or is_cyanosis)
                else "Patient presents visually calm without signs of acute physiological strain."
            )
        }
    ]

    red_flags = []
    recs = []

    stroke_risk = "Low"
    distress_level = "Moderate"
    score = pain_scale

    if is_drooping:
        stroke_risk = "High"
        distress_level = "Critical"
        score = max(score, 9)
        red_flags.append("🚨 FAST ALERT: Facial Asymmetry / Unilateral Muscle Droop Detected")
        recs.append("Initiate Immediate FAST Stroke Protocol")
        recs.append("Order Emergency Non-contrast Head CT")
        recs.append("Stat Neurology Consultation")

    if is_cyanosis:
        distress_level = "Critical"
        score = max(score, 9)
        red_flags.append("🚨 HYPOXIA ALERT: Visible Perioral Cyanosis")
        recs.append("Check Pulse Oximetry (SpO2) stat")
        recs.append("Administer Supplemental Oxygen")

    if is_pallor:
        score = max(score, score + 1)
        red_flags.append("⚠️ CLINICAL ALERT: Skin Pallor / Potential Perfusion Deficit")
        recs.append("Assess Hemoglobin & Capillary Refill Time")

    if is_swelling:
        score = max(score, score + 1)
        red_flags.append("⚠️ ALLERGY ALERT: Facial / Periorbital Swelling Detected")
        recs.append("Screen for Airway Compromise / Anaphylaxis")

    if pain_scale >= 8:
        if distress_level != "Critical":
            distress_level = "High"
        red_flags.append("⚠️ High Facial Pain Expression (Severe Grimacing)")
        recs.append("Initiate Acute Pain Protocol")

    if not recs:
        recs.append("Visual Assessment Completed")
        recs.append("Proceed with Standard Vital Signs Triage")

    expression = (
        "Severe Facial Grimacing & Tension"
        if pain_scale >= 8
        else ("Moderate Facial Distress" if pain_scale >= 5 else "Neutral / Relaxed")
    )

    return {
        "facial_pain_score": min(10, max(1, score)),
        "distress_level": distress_level,
        "stroke_asymmetry_risk": stroke_risk,
        "detected_expression": expression,
        "observations_breakdown": observations_breakdown,
        "red_flags": red_flags,
        "recommendations": list(dict.fromkeys(recs)),
        "confidence": "High",
        "ai_vision_mode": "Local Multi-Observation Vision Engine"
    }


def evaluate_face_image(body: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    body = body or {}
    image_base64 = body.get("image_base64")

    if not ANTHROPIC_API_KEY or not image_base64:
        return mock_face_analysis_fallback(body)

    try:
        clean_base64 = str(image_base64)
        if "base64," in clean_base64:
            clean_base64 = clean_base64.split("base64,")[1]

        client = anthropic.Anthropic(api_key=ANTHROPIC_API_KEY)

        user_prompt = (
            "Analyze this patient facial photo for ER triage decision support across 8 clinical observations:\n"
            "1. Pain expression (1-10 severity)\n"
            "2. Facial asymmetry / unilateral drooping (FAST stroke sign)\n"
            "3. Eye abnormalities (ptosis, scleral redness, pupil size)\n"
            "4. Swelling / edema (facial or lip swelling)\n"
            "5. Skin pallor (paleness)\n"
            "6. Cyanosis (bluish skin tint)\n"
            "7. Fatigue / lethargy\n"
            "8. Visible acute distress\n\n"
            "Provide JSON with schema:\n"
            "{\n"
            '  "facial_pain_score": <int 1-10>,\n'
            '  "distress_level": "Low" | "Moderate" | "High" | "Critical",\n'
            '  "stroke_asymmetry_risk": "Low" | "Medium" | "High",\n'
            '  "detected_expression": "<string>",\n'
            '  "observations_breakdown": [\n'
            '    { "name": "<Observation Name>", "status": "<string>", "severity": "Low"|"Moderate"|"High"|"Critical", "explanation": "<detailed explanation>" }\n'
            "  ],\n"
            '  "red_flags": [<string>, ...],\n'
            '  "recommendations": [<string>, ...],\n'
            '  "confidence": "High" | "Medium" | "Low"\n'
            "}"
        )

        response = client.messages.create(
            model=CLAUDE_MODEL,
            max_tokens=700,
            messages=[
                {
                    "role": "user",
                    "content": [
                        {
                            "type": "image",
                            "source": {
                                "type": "base64",
                                "media_type": "image/jpeg",
                                "data": clean_base64
                            }
                        },
                        {
                            "type": "text",
                            "text": user_prompt
                        }
                    ]
                }
            ]
        )

        text = response.content[0].text
        start_idx = text.find("{")
        end_idx = text.rfind("}")
        if start_idx != -1 and end_idx != -1:
            data = json.loads(text[start_idx : end_idx + 1])
            data["ai_vision_mode"] = "Claude 3.5 Vision AI (Multi-Observation)"
            return data

        return mock_face_analysis_fallback(body)

    except Exception as e:
        logger.warning(f"Claude Vision API call failed: {e}. Falling back to local engine.")
        return mock_face_analysis_fallback(body)
